File: BigWorldServer/world/MoveTransitionTable.go
package main

import (
	"fmt"
	"os"

	pb "bigworld/common/pb"
)

type MoveTransitionTable struct {
	allowed map[pb.MoveState]map[pb.MoveState]bool
}

var moveTransitionTable = &MoveTransitionTable{}

func MoveTransitions() *MoveTransitionTable { return moveTransitionTable }

func (t *MoveTransitionTable) Load(path string) error {
	data, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	msg := pb.GetRootAsTransitionTableMsg(data, 0)
	t.allowed = make(map[pb.MoveState]map[pb.MoveState]bool)
	e := &pb.TransitionEntry{}
	for i := 0; i < msg.EntriesLength(); i++ {
		if !msg.Entries(e, i) {
			continue
		}
		from, ok := StateNameToMoveState(string(e.Source()))
		if !ok {
			return fmt.Errorf("state transition table: unknown source state %q", e.Source())
		}
		targets := make(map[pb.MoveState]bool)
		for j := 0; j < e.AllowedTargetsLength(); j++ {
			to, ok := StateNameToMoveState(string(e.AllowedTargets(j)))
			if !ok {
				return fmt.Errorf("state transition table: unknown target state %q for source %q", e.AllowedTargets(j), e.Source())
			}
			targets[to] = true
		}
		t.allowed[from] = targets
	}
	if len(t.allowed) == 0 {
		return fmt.Errorf("state transition table has no usable entries")
	}
	return nil
}

func (t *MoveTransitionTable) CanTransition(from, to pb.MoveState) bool {
	if from == to {
		return true
	}
	if len(t.allowed) == 0 {
		return true
	}
	targets, ok := t.allowed[from]
	if !ok {
		return false
	}
	return targets[to]
}

File: BigWorldClient/Assets/Config/MoveTransitionTable.cs
using System.Collections.Generic;
using BigWorldClient.Network.Protocol;
using UnityEngine;

namespace BigWorldClient
{

    public sealed class MoveTransitionTable
    {
        private readonly Dictionary<MoveState, HashSet<MoveState>> _allowed = new();

        static MoveTransitionTable _cached;

        public static MoveTransitionTable Instance => _cached ??= Load();

        static MoveTransitionTable Load()
        {
            var table = new MoveTransitionTable();
            TextAsset asset = Resources.Load<TextAsset>("Config/StateTransitionTable");
            if (asset == null) return table;

            TransitionTableMsg msg = TransitionTableMsg.GetRootAsTransitionTableMsg(
                new Google.FlatBuffers.ByteBuffer(asset.bytes));
            for (int i = 0; i < msg.EntriesLength; i++)
            {
                var entry = msg.Entries(i);
                if (!entry.HasValue || !TryParseStateName(entry.Value.Source, out MoveState from)) continue;

                HashSet<MoveState> targets = new();
                for (int j = 0; j < entry.Value.AllowedTargetsLength; j++)
                {
                    if (TryParseStateName(entry.Value.AllowedTargets(j), out MoveState to))
                        targets.Add(to);
                }
                table._allowed[from] = targets;
            }
            return table;
        }

        public bool CanTransition(MoveState from, MoveState to)
        {
            if (from == to) return true;
            if (_allowed.Count == 0) return true;
            return _allowed.TryGetValue(from, out HashSet<MoveState> targets) && targets.Contains(to);
        }

        private static bool TryParseStateName(string name, out MoveState state)
        {
            var mappings = MoveStateMappings.Mappings;
            for (int i = 0; i < mappings.Length; i++)
            {
                if (mappings[i].Name == name)
                {
                    state = mappings[i].State;
                    return true;
                }
            }

            state = MoveState.MoveIdle;
            return false;
        }
    }
}
